### PR TITLE
server: fix command queue endpoint

### LINKED ISSUE #21239: ui/debug: Range debug page should link to the correct CommandQueue viz page
The range debug page attempts to only show a link to the leader's CommandQueue viz page. Unfortunately, clicking on this link brings us to the current gateway's CommandQueue viz page, which may not be the leader and may not even contain a replica in the range. The former will show an empty CommandQueue (almost always, although it is technically possible for a follower to have commands in its CommandQueue) and the latter will throw an error (see https://github.com/cockroachdb/cockroach/issues/21238). This should be improved. There are two things we can do here:
1. The gRPC-gateway endpoint could internally redirect to the leaseholder's node which would in-turn look for the `Range` in its `Store`. This would allow all admin UI gateways to be useful, not just the leaseholder's admin UI gateway.
2. The link could redirect to the correct admin UI gateway. 

The first option is probably better as it fits with how this kind of thing is done in other places. The only downside is that it may be tricky to guarantee that we'll end up redirecting to and reading from the `CommandQueue` of the correct `Replica`.

### PR DESCRIPTION
**Pull request by vilterp**
Previously, the endpoint would return an error if the node didn't have the leaseholder replica for the requested range. Now, it will fan out the request to all nodes, and return the response from the node which had the leaseholder range.

Fixes #21239. The suggestion there was to link to the correct node, but that's hard because we don't know its exposed address. This should achieve the same result.

Release note: None

### CODE REVIEW

File: pkg/server/status_test.go [hunk 1]
```diff
@@ -985,3 +985,25 @@ func TestStatusAPIStatements(t *testing.T) {
...
+
+func TestStatusAPICommandQueue(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	numNodes := 3
+	testCluster := serverutils.StartTestCluster(t, numNodes, base.TestClusterArgs{})
+	defer testCluster.Stopper().Stop(context.Background())
+
+	// Verify that we can get the command queue without error from any node, and that
+	// the response won't be empty.
```
Comment (vilterp) [inline pkg/server/status_test.go]: This test still needs a bit of work — seems to pass without the fix. Probably need to increase cluster size and/or hit a different range to repro.

**Review by a-robinson — CHANGES_REQUESTED** [review]
I'm gonna be out next week, so you may want to have someone else finish the review if you don't want to wait. Sorry!

> Reviewed 4 of 5 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31846#-:-LPih-OY-5aEi-X0RDOu:b-uyk8vf)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/server/status.go, line 1331 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31846#-LPiZDUPCb9ZgpjVuOFN:-LPiZDUPCb9ZgpjVuOFO:b-yxn6ao) ([raw file](https://github.com/cockroachdb/cockroach/blob/97ae2dde543f64842e298867e83ca6555bcb2e5e/pkg/server/status.go#L1331)):*
> ```Go
> 	ctx context.Context, req *serverpb.CommandQueueRequest,
> ) (*serverpb.CommandQueueResponse, error) {
> 	switch req.NodeID {
> ```

It's a more common/preferred style to, when possible, handle the simpler cases first and return instead of nesting everything in a large switch statement. That fewer levels of indentation wrapping a large block of code, the better.
___
*[pkg/server/status.go, line 1336 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31846#-LPienIgAP0H-tlg4TKR:-LPienIgAP0H-tlg4TKS:bygrhlp) ([raw file](https://github.com/cockroachdb/cockroach/blob/97ae2dde543f64842e298867e83ca6555bcb2e5e/pkg/server/status.go#L1336)):*
> ```Go
> 		// that response.
> 		overallResp := &serverpb.CommandQueueResponse{}
> 		dialFun := func(ctx context.Context, nodeID roachpb.NodeID) (i interface{}, e error) {
> ```

s/Fun/Fn/g

For whatever reason it's the idiom we use everywhere.
___
*[pkg/server/status.go, line 1349 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31846#-LPigaKqC6OtdaulfxOf:-LPigaKqC6OtdaulfxOg:be7fnhm) ([raw file](https://github.com/cockroachdb/cockroach/blob/97ae2dde543f64842e298867e83ca6555bcb2e5e/pkg/server/status.go#L1349)):*
> ```Go
> 		}
> 		errorFun := func(nodeID roachpb.NodeID, nodeFnError error) {
> 			log.Warningf(ctx, "error getting command queue for node %d: %+v", nodeID, nodeFnError)
> ```

This is gonna log way too much given that looks like it'll be called on every node that isn't the leaseholder. I'd only log if it hits an error that isn't a `RangeNotFoundError` or `NotLeaseHolderError`.

We may also want a plan for what to return if none of the nodes returns success.
___
*[pkg/server/status.go, line 1389 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31846#-LPigHsv9JAG8iYFdzGA:-LPigHsw9Ml8XEoLdxA4:b-95enh4) ([raw file](https://github.com/cockroachdb/cockroach/blob/97ae2dde543f64842e298867e83ca6555bcb2e5e/pkg/server/status.go#L1389)):*
<details><summary><i alt="summary">Quoted 5 lines of code…</i></summary>

> ```
> 	timestamp := s.admin.server.clock.Now()
> 	isLiveMap := s.nodeLiveness.GetIsLiveMap()
> 	availableNodes := s.storePool.AvailableNodeCount()
> 	metrics := replica.Metrics(ctx, timestamp, isLiveMap, availableNodes)
> 	if !metrics.Leaseholder {
> ```
</details>


Replace all this with `replica.OwnsValidLease(s.admin.server.clock.Now()`.
___
*[pkg/server/status_test.go, line 997 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31846#-LPdWIHS9RX7Fj9ZuL7v-r1-997:-LPigslF4oi3SbE0PWCr:b-bmet9m) ([raw file](https://github.com/cockroachdb/cockroach/blob/97ae2dde543f64842e298867e83ca6555bcb2e5e/pkg/server/status_test.go#L997)):*
<details><summary><i>Previously, vilterp (Pete Vilter) wrote…</i></summary><blockquote>

This test still needs a bit of work — seems to pass without the fix. Probably need to increase cluster size and/or hit a different range to repro.
</blockquote></details>

That is indeed concerning. Worth figuring out before merging this.
___
*[pkg/server/serverpb/status.proto, line 620 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31846#-LPiapsI96FOXmiHeW7L:-LPiapsI96FOXmiHeW7M:bbedfmd) ([raw file](https://github.com/cockroachdb/cockroach/blob/97ae2dde543f64842e298867e83ca6555bcb2e5e/pkg/server/serverpb/status.proto#L620)):*
> ```protobuf
>   // - If this is not the current node, it will be forwarded.
>   // - If this is 0, it will be forwarded to all nodes.
>   int32 node_id = 2 [
> ```

Just a heads up that using an int32 instead of a string for this is actually going against the grain in this API. Although I think that all the uses that are strings default to just the local node, so maybe differentiating is better. This is totally up to you, I'm just calling it out in case you didn't notice.
___
*[pkg/server/serverpb/status.proto, line 627 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31846#-LPickCbB2yYmNwWFv1k:-LPickCbB2yYmNwWFv1l:b-xmkew6) ([raw file](https://github.com/cockroachdb/cockroach/blob/97ae2dde543f64842e298867e83ca6555bcb2e5e/pkg/server/serverpb/status.proto#L627)):*
> ```protobuf
> 
> message CommandQueueResponse {
>   storage.storagepb.CommandQueuesSnapshot snapshot = 1
> ```

If it the result may be coming from a different node than the local node, we should probably include the nodeID in the response.

**Review by vilterp — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31846)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/server/status.go, line 1331 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31846#-LPiZDUPCb9ZgpjVuOFN:-LQ0_dFg7pLyKzOhTV0A:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/97ae2dde543f64842e298867e83ca6555bcb2e5e/pkg/server/status.go#L1331)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

It's a more common/preferred style to, when possible, handle the simpler cases first and return instead of nesting everything in a large switch statement. That fewer levels of indentation wrapping a large block of code, the better.
</blockquote></details>

Done.
___
*[pkg/server/status.go, line 1336 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31846#-LPienIgAP0H-tlg4TKR:-LQ0_cqj2KWUzspe-H-Z:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/97ae2dde543f64842e298867e83ca6555bcb2e5e/pkg/server/status.go#L1336)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

s/Fun/Fn/g

For whatever reason it's the idiom we use everywhere.
</blockquote></details>

Done.
___
*[pkg/server/status.go, line 1349 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31846#-LPigaKqC6OtdaulfxOf:-LQ10bORBofm3l_k9v0-:b-un68s6) ([raw file](https://github.com/cockroachdb/cockroach/blob/97ae2dde543f64842e298867e83ca6555bcb2e5e/pkg/server/status.go#L1349)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

This is gonna log way too much given that looks like it'll be called on every node that isn't the leaseholder. I'd only log if it hits an error that isn't a `RangeNotFoundError` or `NotLeaseHolderError`.

We may also want a plan for what to return if none of the nodes returns success.
</blockquote></details>

Hm, can't check the type of the error since it's wrapped in an `errors.withStack`. Changed it to just do nothing; maybe better to capture it in a variable at the level of the endpoint function, and return that?

Changed the endpoint to return an error if none of them are successful.
___
*[pkg/server/status.go, line 1389 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31846#-LPigHsv9JAG8iYFdzGA:-LQ0_ylCE6gWGbWCzXiQ:b-dxtqxv) ([raw file](https://github.com/cockroachdb/cockroach/blob/97ae2dde543f64842e298867e83ca6555bcb2e5e/pkg/server/status.go#L1389)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

> ```
> 	timestamp := s.admin.server.clock.Now()
> 	isLiveMap := s.nodeLiveness.GetIsLiveMap()
> 	availableNodes := s.storePool.AvailableNodeCount()
> 	metrics := replica.Metrics(ctx, timestamp, isLiveMap, availableNodes)
> 	if !metrics.Leaseholder {
> ```

Replace all this with `replica.OwnsValidLease(s.admin.server.clock.Now()`.
</blockquote></details>

Done. Perfect, thanks.
___
*[pkg/server/status_test.go, line 997 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31846#-LPdWIHS9RX7Fj9ZuL7v-r1-997:-LQ16du56dJO1LZqj6RC:b-tydawb) ([raw file](https://github.com/cockroachdb/cockroach/blob/97ae2dde543f64842e298867e83ca6555bcb2e5e/pkg/server/status_test.go#L997)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

That is indeed concerning. Worth figuring out before merging this.
</blockquote></details>

Done. Needed to increase the test cluster size beyond the system replication factor.
___
*[pkg/server/serverpb/status.proto, line 620 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31846#-LPiapsI96FOXmiHeW7L:-LQ1-JTR8aJ9fEl2u08u:b-ihtmeg) ([raw file](https://github.com/cockroachdb/cockroach/blob/97ae2dde543f64842e298867e83ca6555bcb2e5e/pkg/server/serverpb/status.proto#L620)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Just a heads up that using an int32 instead of a string for this is actually going against the grain in this API. Although I think that all the uses that are strings default to just the local node, so maybe differentiating is better. This is totally up to you, I'm just calling it out in case you didn't notice.
</blockquote></details>

I indeed didn't notice :) Changed to take a string, and documented the what each value means. This defaults to fanout, and allows specifying "local", like the others. Seems like a decent compromise.
___
*[pkg/server/serverpb/status.proto, line 627 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31846#-LPickCbB2yYmNwWFv1k:-LQ1-icr4pgBPKFeCFKQ:bncy0mz) ([raw file](https://github.com/cockroachdb/cockroach/blob/97ae2dde543f64842e298867e83ca6555bcb2e5e/pkg/server/serverpb/status.proto#L627)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

If it the result may be coming from a different node than the local node, we should probably include the nodeID in the response.
</blockquote></details>

I guess that could help in the local case, but we'd have to return multiple node ids in the fanout case, and just pass you back the node id you passed in in the explicit node id case. Not really seeing the value of including this.

**Comment by vilterp** [comment]
Not sure how to fix this test flake:

```
soon.go:49: condition failed to evaluate within 45s: Get https://127.0.0.1:33803/_status/range/1/cmdqueue: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
      goroutine 54 [running]:
      runtime/debug.Stack(0xa7a358200, 0xc428bb5d10, 0x3987fe0)
        /usr/local/go/src/runtime/debug/stack.go:24 +0xb5
      github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon(0x39e2080, 0xc420275590, 0xc4252182a0)
        /go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:50 +0x172
      github.com/cockroachdb/cockroach/pkg/server.TestStatusAPICommandQueue(0xc420275590)
        /go/src/github.com/cockroachdb/cockroach/pkg/server/status_test.go:1005 +0x1f1
      testing.tRunner(0xc420275590, 0x32bec98)
        /usr/local/go/src/testing/testing.go:777 +0x16e
      created by testing.(*T).Run
        /usr/local/go/src/testing/testing.go:824 +0x565
```

@a-robinson any ideas?

**Review by a-robinson — CHANGES_REQUESTED** [review]
No ideas, there's not a whole lot to go off there. Have you seen it any more than just the one time?

> Reviewed 4 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31846#-:-LREJdy84299Pms7gZ8B:b-ar5faa)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/server/status.go, line 1349 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31846#-LPigaKqC6OtdaulfxOf:-LREIK2QC_GLIfUSz6_F:b-rxmyh8) ([raw file](https://github.com/cockroachdb/cockroach/blob/97ae2dde543f64842e298867e83ca6555bcb2e5e/pkg/server/status.go#L1349)):*
<details><summary><i>Previously, vilterp (Pete Vilter) wrote…</i></summary><blockquote>

Hm, can't check the type of the error since it's wrapped in an `errors.withStack`. Changed it to just do nothing; maybe better to capture it in a variable at the level of the endpoint function, and return that?

Changed the endpoint to return an error if none of them are successful.
</blockquote></details>

Does `errors.Cause` not work for extracting the original error?

If we can't reasonably get to the root error, then I'd suggest switching this to a `log.VEventf(ctx, 1, ...)` so we can at least get at the issue if/when we really need to.
___
*[pkg/server/status.go, line 1331 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/31846#-LREHGWc01uxcvb72bch:-LREHGWd13mayLpzSX4J:b2wei1o) ([raw file](https://github.com/cockroachdb/cockroach/blob/7777977eb17e8f266de7004c8d895476da477d14/pkg/server/status.go#L1331)):*
> ```Go
> 	ctx context.Context, req *serverpb.CommandQueueRequest,
> ) (*serverpb.CommandQueueResponse, error) {
> 	if req.NodeID == "local" || req.NodeID == strconv.Itoa(int(s.admin.server.NodeID())) {
> ```

Just use `s.parseNodeID()` here for parsing and determination of whether the request is local like all the other functions in this file do.
___
*[pkg/server/status.go, line 1345 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/31846#-LREHiI4CitaEbKDtzsD:-LREHiI4CitaEbKDtzsE:b-hgt7gx) ([raw file](https://github.com/cockroachdb/cockroach/blob/7777977eb17e8f266de7004c8d895476da477d14/pkg/server/status.go#L1345)):*
> ```Go
> 			return nil, err
> 		}
> 		return statusClient.CommandQueue(ctx, &serverpb.CommandQueueRequest{
> ```

Why not just pass on the provided `req`?

**Comment by vilterp** [comment]
Command queue page (and this endpoint) are deleted in #32865

**Pull request closed without merging**